### DISCUSSION TITLE
Some questions about latent variables in Probabilistic Diﬀeomorphic Registration

### DISCUSSION THREAD

**Discussion by cs123951** [voxelmorph/voxelmorph#416] [Q&A]
Hello,
I have read your paper:
A. V. Dalca, G. Balakrishnan, J. Guttag, and M. R. Sabuncu, “Unsupervised Learning for Fast Probabilistic Diffeomorphic Registration,” in Medical Image Computing and Computer Assisted Intervention – MICCAI 2018, vol. 11070, Cham: Springer International Publishing, 2018, pp. 729–738. doi: [10.1007/978-3-030-00928-1_82](https://doi.org/10.1007/978-3-030-00928-1_82).

I could not figure out why the velocity field z can be used. As said in this paper:

> In this paper, we let z be a stationary velocity ﬁeld that speciﬁes a diﬀeomorphism through the ODE (1).

> We ﬁrst implement a layer that samples a new z k∼ N( μ zx,y| , Σ zx,y| ) using the “re-parameterization trick”

I have learned about re-parameterization trick. The authors who proposed re-parameterization trick used a vector to represent Z.
I think that is because the components of Z are independent of each other, isn't it? 

But you used a matrix (the velocity field) to represent Z. Then you used convolution operations to process Z. Once convolution operation is used, it is equivalent to assuming that the components of Z have a local correlation with each other. That is conflicting with the assumption that the components of Z are independent of each other.

I found that many works used a vector to represent Z. It is creative to use a matrix as Z and then use a convolution network. But I want to know why this can be done? What is the difference with a vector Z?

Thank you if you could answer my question!

**Comment by adalca** [comment]
> I think that is because the components of Z are independent of each other, isn't it?

I'm not sure what you mean by that. 

The re-parametrization "trick" is just sampling math for a gaussian. To sample a gaussian z ~ N(u, s) you can simply sample from r from N(0, 1) and form z = rs + u, which will ensure z samples have mean u and variance s. 

> Then you used convolution operations to process Z. 

Can you explain where this is or what you mean? Where to we use a convolution after sampling Z?

> I found that many works used a vector to represent Z. It is creative to use a matrix as Z and then use a convolution network. But I want to know why this can be done? What is the difference with a vector Z?

I'm not quite sure what you mean by matrix vs vector here? Z is just a vector that is reshaped as the volume size, conceptually

**Reply by cs123951** [reply]
I may not describe it properly. When I said matrix in the above discussion, I intend to mean tensor which has n dimensions. 
A k-by-1 matrix is called a column vector and a 1-by-k matrix is called a row vector. 

As you said in the paper:
> We estimate μ and Σ using a convolutional neural network def ψ(x, y ) parameterized by ψ, as described below.
> We design the network def ψ (x, y ) that takes as input x and y and outputs μ and Σ, based on a 3D UNet-style architecture

The size of the input 3D image (i.e., x or y) is 160×192×224. The network is a U-net. So the U-net outputs a tensor of 160×192×224x nf,  nf is the number of filters in the last step of U-net. Let's call this tensor unet_out.
Then how do you get μ and Σ from unet_out? 

There are two ways:
(1) Add a convolution layer after unet_out, that is: 
```
vt_mean = vae_mean(unet_out_flattened)  # vae_mean = Convolution_mean()
vt_logvar = vae_logvar(unet_out_flattened)  # vae_logvar = Convolution_logvar()
```
Then we can sample epsilon, and obtain z by z=exp(0.5*vt_logvar)*epsilon+vt_mean

I used to think that you have adopted this method, so my questions are all about this kind of situation.
But you said,

> Z is just a vector that is reshaped as the volume size, conceptually

I think I may have gone in the wrong direction. You may use the second method.

(2)Add a fully connected layer after unet_out, that is
```
unet_out_flattened = unet_out.view(batch, flatten_size) 
vt_mean = vae_mean(unet_out_flattened)  # vae_mean = Linear_mean(flatten_size, flatten_size)
vt_logvar = vae_logvar(unet_out_flattened)  # vae_logvar = Linear_logvar(flatten_size, flatten_size)
```
But this fully connected layer cost too much memory. For example, if z is the displacement ﬁeld itself, then the number of weights in the fully connected layer is (160×192×224)^2=47352014438400. If z is a low-dimensional embedding of a dense displacement ﬁeld, then how did you get the low-dimensional embedding? By fully connected layer or convolutional layer?

How do you deal with large memory consumption?
Is it possible to add another layer after z?

**Comment by adalca** [comment]
Sorry I didn't mean that we do a dense layer or anything like that, I just meant that the _matrix_ vs _vector_ definitions are an unnecessary distinction and part of the confusion in this situation. 

Definitions aside, we do something like what you describe in "1)".

The mu and logvar are generated with a convnet, but the z is sampled from N(mu, var), which means that each element on z is independent. 

in the original paper we don't do any convolutions after this (after sampling z), so the components of z are independently sampled, that is z_i ~ N(mu_i, var_i)

We had a medneurips follow-up paper where we showed that you can do convolutions on the sampled tensor, which is equivalent to adding structure to the covariance, but that's not in the MICCAI2018 paper.

**Reply by cs123951** [reply]
Thank you!
My questions with the MICCAI2018 paper have fully been solved. That's really excellent work!

By the way, I have found the follow-up paper titled  "Improved Probabilistic Diffeomorphic Registration with CNNs".
Reading 
<img width="829" alt="image" src="https://user-images.githubusercontent.com/13451757/162247081-e884885e-4da2-437c-8c21-a050a8c6469c.png">

C can be regarded as a  3D convolution of Dr. 
The elements in r are independent because r is a sample from the standard normal. I deduce that the elements in Dr are independent, aren't they?
Convolutions are often used for processing data that has a known, grid-like topology. For example, convolution is often used in images while the pixels are related to their neighbors.  The idea behind CDr is that the elements in Dr have a relationship so that the convolution operation can work.  But the elements in Dr are independent. **Isn't this contradictory?**

My deduction can be concluded in 3 steps:
1. The elements in Dr are independent.
2. The w in Cw has to be dependent.
3. So C can not impose on Dr.   C Dr can not work.

**Comment by adalca** [comment]
I'm not sure I understand your reasoning here, especially for steps 2 and 3. `Dr` has independent elements, and CDr are spatialy dependent (thus the resulting z has spatial dependencies). I'm not sure where the contradiction is?

Here's one way to understand it: the miccai 2018 modelling assumed z is sampled from a gaussian with a simple variance. This is a modelling assumption, and it's maybe  a bit unnatural since in reality voxels are not independent.

So in the follow up we said we'd assume a *covariance* structure. But sampling z from N(mu, Sigma) where Sigma is very complex is impractical (since Sigma is huge). Therefore we came up with an approximation to Sigma that is practical *and* offers some covariance -- whish is Sigma = CDD'C'. Here, D is diagonal and C is a spatial convolution operator.  This, again, is a *modelling* choice, one that is more powerful than the MICCAI2018 (and maybe more realistic), but is still practical in our sampling strategy.

**Reply by cs123951** [reply]
> This is a modeling assumption, and it's maybe a bit unnatural since in reality voxels are not independent.

Yes, this is the crux of the matter. Now that the assumption ( z is independent ) does not belong to the subset of reality, how can this model be used in practical applications? Or what is the point of the model?

**Comment by adalca** [comment]
>  how can this model be used in practical applications?

All models are wrong (there's even a wikipedia article about it!: https://en.wikipedia.org/wiki/All_models_are_wrong). 

More concretely, all models make approximating assumptions to make them tractable. In our case, the variational approximation enables reasonable (plausibly computable) inference (registration). By assuming z is drawn from a gaussian, we still capture most of the spatial relationship (in the mean of the gaussian), but believe the *noise* is independent (which is still a strong assumption, but not that crazy). 
As you can see in the results of the papers, this is a reasonable tradeoff in practice, giving good registration results.

**Reply by cs123951** [reply]
Thank you! I have fully solved my problems! I am so inspired by your reply. 
Wish you every success in your career.
